lib/HLSL/DxilCondenseResources.cpp

```diff
@@ -2781,12 +2783,329 @@ ModulePass *llvm::createDxilAllocateResourcesForLibPass() {
+      annotHdls.emplace_back(CI);
+    }
+  }
+  if (annotHdls.empty())
+    return false;
+
+  DenseSet<Instruction *> selectAnnotHdls;
+  for (CallInst *CI : annotHdls) {
+    for (User *U : CI->users()) {
+      if (isa<PHINode>(U) || isa<SelectInst>(U))
+        selectAnnotHdls.insert(cast<Instruction>(U));
+    }
+  }
+  const ShaderModel *pSM = DM.GetShaderModel();
+  Type *propsTy = op->GetResourcePropertiesType();
+  Value *OpArg =
+      op->GetI32Const(static_cast<unsigned>(DXIL::OpCode::AnnotateHandle));
+  // Sink annotateHandle after phi.
+  for (Instruction *Hdl : selectAnnotHdls) {
+    if (PHINode *phi = dyn_cast<PHINode>(Hdl)) {
+      Value *props = nullptr;
+      for (unsigned i = 0; i < phi->getNumIncomingValues(); ++i) {
+        Value *V = phi->getIncomingValue(i);
+        if (CallInst *CI = dyn_cast<CallInst>(V)) {
+          DxilInst_AnnotateHandle annot(CI);
+          if (annot) {
+            if (props == nullptr) {
+              props = annot.get_props();
+            }
+            else if (props != annot.get_props()) {
+              props = resource_helper::tryMergeProps(
+                  cast<Constant>(props), cast<Constant>(annot.get_props()),
+                  propsTy, *pSM);
+              if (props == nullptr) {
+                DXASSERT(false, "mismatch handle annotation");
+              }
+            }
+
+            Value *res = annot.get_res();
+            phi->setIncomingValue(i, res);
+          }
+        }
+      }
+      // Insert after phi.
+      IRBuilder<> B(phi->getParent()->getFirstNonPHI());
+      CallInst *annotCI = B.CreateCall(annotHdlFn, {OpArg, phi, props});
+      phi->replaceAllUsesWith(annotCI);
+      annotCI->setArgOperand(DxilInst_AnnotateHandle::arg_res, phi);
+    } else {
+      SelectInst *sel = dyn_cast<SelectInst>(Hdl);
+      Value *TVal = sel->getTrueValue();
+      Value *FVal = sel->getFalseValue();
+      Value *props = nullptr;
+      if (CallInst *CI = dyn_cast<CallInst>(TVal)) {
+        DxilInst_AnnotateHandle annot(CI);
+        if (annot) {
+          props = annot.get_props();
+          Value *res = annot.get_res();
+          sel->setOperand(1, res);
+        }
+      }
+      if (CallInst *CI = dyn_cast<CallInst>(FVal)) {
+        DxilInst_AnnotateHandle annot(CI);
+        if (annot) {
+          if (props == nullptr) {
+            props = annot.get_props();
+          } else if (props != annot.get_props()) {
+            props = resource_helper::tryMergeProps(
+                cast<Constant>(props), cast<Constant>(annot.get_props()),
+                propsTy, *pSM);
+            if (props == nullptr) {
+              DXASSERT(false, "mismatch handle annotation");
+            }
+          }
+
+          Value *res = annot.get_res();
+          sel->setOperand(2, res);
+        }
+      }
+
+      // Insert after sel.
+      IRBuilder<> B(sel->getNextNode());
+      CallInst *annotCI = B.CreateCall(annotHdlFn, {OpArg, sel, props});
```

Comment:
Initial value for resource here could be undef to prevent replacement of use with self before it's re-set below.